src/Native/Runtime/arm/Interlocked.S

```diff
@@ -13,17 +16,34 @@
 // r2 = comparand
 LEAF_ENTRY RhpLockCmpXchg32, _TEXT
 ALTERNATE_ENTRY RhpLockCmpXchg32AVLocation
-    bx      lr
+LOCAL_LABEL(CmpXchg32Retry):
+          ldrex        r3, [r0]
+          cmp          r2, r3
+          bne          LOCAL_LABEL(CmpXchg32Exit)
+          strex        r3, r1, [r0]
+          cbnz         r3, LOCAL_LABEL(CmpXchg32Retry)
+LOCAL_LABEL(CmpXchg32Exit):
+          bx           lr
 LEAF_END RhpLockCmpXchg32, _TEXT
 
-
 // WARNING: Code in EHHelpers.cpp makes assumptions about this helper, in particular:
-// - Function "InWriteBarrierHelper" assumes an AV due to passed in null pointer will happen at RhpLockCmpXchg64AVLocation
+// - Function "InWriteBarrierHelper" assumes an AV due to passed in null pointer will happen at RhpLockCmpXchg32AVLocation
 // - Function "UnwindWriteBarrierToCaller" assumes the stack contains just the pushed return address
-// r0 = destination address
-// r1 = value
-// r2 = comparand
+// r0      = destination address
+// {r1,r2} = value
+// sp[0+8] = comparand
 LEAF_ENTRY RhpLockCmpXchg64, _TEXT
 ALTERNATE_ENTRY RhpLockCmpXchg64AVLocation
-	bx	lr
+          PROLOG_PUSH  "{r4-r6,lr}"
+          ldmia        sp!, {r5-r6}
+LOCAL_LABEL(CmpXchg64Retry):
+          ldrexd       r3, r4, [r0]
+          eor          r3, r5
+          eor          r4, r6
+          orr          r3, r4
+          cbnz         r3, LOCAL_LABEL(CmpXchg64Exit)
+          strexd       r3, r1, r2, [r0]
+          cbz          r3, LOCAL_LABEL(CmpXchg64Retry)
+LOCAL_LABEL(CmpXchg64Exit):
+          EPILOG_POP   "{r4-r6,pc}"
```

Comment:
The function needs to return the original value of `[r0]`. I do not see anything that does that.